cbits/eventlog_socket.c

```diff
@@ -0,0 +1,531 @@
+    pthread_mutex_unlock(&mutex);
+    return;
+  }
+
+  if (pfd.revents & POLLOUT) {
+    DEBUG_ERR("(%d) POLLOUT\n", fd);
+
+    pthread_mutex_lock(&mutex);
+    while (wt.head) {
+      struct write_buffer_item *item = wt.head;
+      ret = write(client_fd, item->data, item->size);
+
+      if (ret == -1) {
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+          // couldn't write anything, shouldn't happend.
+          // do nothing.
+        } else if (errno == EPIPE) {
+          client_fd = -1;
+          write_buffer_free(&wt);
+        } else {
+          PRINT_ERR("failed to write: %s\n", strerror(errno));
+        }
+
+        // break out of the loop
+        break;
+
+      } else {
+        // we wrote something
+        if (ret >= item->size) {
+          // we wrote whole element, try to write next element too
+          write_buffer_pop(&wt);
+          continue;
+        } else {
+          item->size -= ret;
+          item->data += ret;
+          break;
+        }
+      }
+    }
+    pthread_mutex_unlock(&mutex);
+  }
+}
+
+static void iteration() {
+  pthread_mutex_lock(&mutex);
+  int fd = client_fd;
+  bool empty = wt.head == NULL;
+  DEBUG_ERR("fd = %d, wt.head = %p\n", fd, wt.head);
+  pthread_mutex_unlock(&mutex);
+
+  if (fd != -1) {
+    if (empty) {
+      nonwrite_iteration(fd);
+    } else {
+      write_iteration(fd);
+    }
+  } else {
+    listen_iteration();
+  }
+}
+
+/* Main loop of eventlog-socket own thread:
+ * Currently it is two states:
+ * - either we have connection, then we poll for writes (and drop of connection).
+ * - or we don't have, then we poll for accept.
+ */
+static void *worker(void * _unused)
+{
+  while (true) {
+    iteration();
+  }
+
+  return NULL; // unreachable
+}
+
+/*********************************************************************************
+ * Initialization
+ *********************************************************************************/
+
+static void open_socket(const char *sock_path)
+{
+  DEBUG_ERR("enter: %s\n", sock_path);
+
+  listen_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+
+  struct sockaddr_un local;
+  local.sun_family = AF_UNIX;
+  strncpy(local.sun_path, sock_path, sizeof(local.sun_path) - 1);
+  unlink(sock_path);
+  if (bind(listen_fd, (struct sockaddr *) &local,
+           sizeof(struct sockaddr_un)) == -1) {
+    PRINT_ERR("failed to bind socket %s: %s\n", sock_path, strerror(errno));
+    abort();
+  }
+
+  int ret = pthread_create(&listen_thread, NULL, worker, NULL);
+  if (ret != 0) {
+    PRINT_ERR("failed to spawn thread: %s\n", strerror(ret));
+    abort();
+  }
+}
+
+
+/*********************************************************************************
+ * Public interface
+ *********************************************************************************/
+
+void eventlog_socket_wait(void)
+{
+  pthread_mutex_lock(&mutex);
+  while (client_fd == -1) {
+    assert(pthread_cond_wait(&new_conn_cond, &mutex) == 0);
```

Comment:
Don't we want to call `pthread_cond_wait` outside of an assert so that it gets executed even if NDEBUG is set?